Bump to 4.0.0

domcleal opened the pull request:
Proposing a new release, assuming 4.x due to drop of 1.9 and probable incompatibility of the `sort` helper.

wvanbergen commented:
Any other cleanup, or breaking changes we could do this timeframe?

wvanbergen commented:
E.g., it may make sense to drop support for ActiveRecord 3.2, 4.0, and 4.1, because they are no longer supported: https://github.com/wvanbergen/scoped_search/blob/master/.travis.yml#L23-L25

domcleal commented:
Yes, that's a good idea. I'd also like to remove the autocomplete JS/JQuery helpers too then, they've been deprecated since 3.0.0 in favour of using the assets pipeline - this would simply leave `sort` and the internal `a_link`. I'll update this PR with these changes if that's OK?

wvanbergen commented:
:ship:

domcleal commented:
PR updated to remove all of the old autocomplete helpers, remove 3.2-4.1 support and various version-based conditionals.

wvanbergen commented on CHANGELOG.rdoc:
```diff
@@ -8,6 +8,20 @@ Please add an entry to the "Unreleased changes" section in your pull requests.
+  `params`. Call `params.permit(..)` and pass in the result to safely generate
+  URLs and prevent non-sanitized errors under Rails 5.
+- Auto completion: escape quotes in listed values (#138)
+- Auto completion: remove javascript helpers in favor of asset pipeline
```
Can we give better instructions on what is needed to co vert?

domcleal replied:
Fixed, included the info about the jQuery scopedSearch method.

wvanbergen commented:
One more idea: adopt keyword arguments.
- Where we are currently using option hashes.
- Where we are currently using long argument list (especially when some of them are optional.

domcleal commented:
> One more idea: adopt keyword arguments.

Yes, this would be good, the registration method signature is quite unwieldy. I can only see a couple of likely candidates:

>     Where we are currently using option hashes.

* `Definition::Field.initialize` (or `Definition#define`), handling the registration of new fields for searching

>     Where we are currently using long argument list (especially when some of them are optional.

* `sort` helper

I'll look a bit more closely and file a separate PR.

wvanbergen commented:
Awesome ❤️

domcleal commented:
Rebased, and updated the changelog to mention deprecations for `:in` and `:alias`.

I've not managed to convert `sort` to keyword arguments easily, the method signature doesn't allow it (it's backwards, options are the first argument) so this can be done in a later minor release I think. It'll probably need a replacement method under another name.

wvanbergen commented:
IMHO, for `sort` we should make `html_options` and `url_options` part of `options`, and convert the result to keyword arguments.

It'll be a backwards incompatible change, but I suspect not many people are using those final 2 argument. Even if you do, your code will blow up in an easy to understand way, and the fix is simple.

domcleal commented:
> IMHO, for sort we should make html_options and url_options part of options, and convert the result to keyword arguments.

Yes, I agree. I was unsure about making such a backwards incompatible change though, but perhaps you're right, it's rare that it's used (I thought Foreman used it heavily, but it doesn't.). It'd be good to get in before `url_options` is commonly used. I'll open a PR for that change.

domcleal commented:
Rebased with the changes to sort() documented.

wvanbergen commented:
🎉

domcleal commented:
Thanks for all of the feedback @wvanbergen. I'll leave you to tag and release the gem.

wvanbergen commented:
Let's make sure you can do that as well. `rake release` should do the trick. If I haven't added you to rubygems.org as owner yet, let me know.

domcleal commented:
You'll need to add me then, dominic at cleal dot org is my registered e-mail address.

wvanbergen commented:
Added!

domcleal commented:
Success, thanks! 4.0.0 is now available.
